uPy/html.py

```diff
@@ -0,0 +1,255 @@
+        val=tag[0:j].strip()
+        tag = tag[j+1:len(tag)]
+        #print("val={0}\ttag={1}".format(val,tag))
+        contents_list.append(val)
+
+    return contents_list
+
+
+
+def construct_form_from_tag_internals(tags):
+    
+    formOpen = False #indicates if <form ...> seen but </form> not seen yet
+    form = {}
+    #From here on we have a list of the contents of a tag.
+    #Decide how to handle it
+
+    for tag_internals in tags: # the splitup tag internals of a single tag
+
+        #form tag: beginning of form
+        if(tag_internals[0]=="form"):
+
+            # Create a new dictionary to store info
+            tagDict = {}
+            tagDict["tag_type"]=tag_internals[0]
+            #print("tagDict[tag_type]= form")
+
+            # insert key="value" pairs
+            j=1
+            while(j<len(tag_internals)):
+                key=tag_internals[j]
+                val=tag_internals[j+1]
+                j+=2
+                tagDict[key] = val
+                #print("\ttagDict[{0}]={1}".format(key,val))
+            
+            # other tags will be inside forms
+            tagDict["inside"] = []
+            form = tagDict
+            formOpen = True
+
+        # /form tag : end of form
+        elif(tag_internals[0]=="/form"):
+            #print("\ttag_type=/form")
+            formOpen=False #close out the form
+
+
+        # a tag: indicates hyperlinked text
+        elif(tag_internals[0]=="a" or tag_internals[0]=="input"):
+            # Create a new dictionary to store info
+            tagDict={}
+            tagDict["tag_type"]=tag_internals[0]
+            #print("\ttagDict[\"tag_type\"]={0}".format(tag_internals[0]))
+
+            # insert key="value" pairs
+            j=1
+            while(j<len(tag_internals)):
+                key=tag_internals[j]
+                val=tag_internals[j+1]
+                j+=2
+                tagDict[key]=val
+                #print("\ttagDict[\"{0}\"]={1}".format(key,val))
+            
+            #append this tag into the form
+            if(formOpen):
+                    inside_list=form["inside"]   #get
+                    inside_list.append(tagDict)  #append
+                    form["inside"]=inside_list   #put back
+
+    return form
+
+
+# TOP LEVEL FUNCTION
+# Input: HTML splashpage as bytes
+# Returns a list of dictionaries. Each dictionary is a form:    
+#   Pick a form by indexing into the list. Each form is a dictionary.   
+#   Dictionary key value pairs are keys and values within that tag line:
+#        ex from html: name="formname" method="POST"
+#        becomes:   dict["name"] -> "formname"
+#                   dict["method"] -> "POST"
+def get_forms(html):
+    
+    html = html.decode('utf-8') # int to str
+
+    # each entry in list is "<form ...> <...> </form>" 
+    forms = get_tags(html, "form")
+    if(len(forms) == 0):
+        return []
+    elif(len(forms) == 1): #caused if <form...> but bad html received no </form>
```

Comment:
I still do not think we should be trying to fix splashpages bad html, we should assume its written correctly and anyone who has bad html is in the 50% we don't break, I'll let you ask the others what they think.

Reply:
This also is a very bad conditional...most pages only have 1 form so checking for this does not mean that there was an incomplete form

Reply:
It does mean there was an incomplete form. This variable is a list of all tags from <form ...> through </form> inclusive.
If the length is zero, then neither <form ...> or </form> were found.
If the length is 1 then <form ...> was found but </form> was not so the find all tags function automatically catches this and ends the tag on > instead of </form>, This works for any tag of the format <x ...> </x>. It is best to not generate an error but rather catch smoothly and recover in a way that can be checked.

tldr, the code is correct as it represents the number of tags from <x ... > to </x> inclusive. So the length from <form ... > without a </form> is just the starting <form ...>, which is one.

Reply:
this is good